File: app/src/main/kotlin/com/spacex/ui/rocketList/RocketListActivity.java
package com.spacex.ui.rocketList;

import static com.spacex.ui.AndroidOSKt.isAtLeastLollipop;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import androidx.annotation.NonNull;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.core.content.ContextCompat;
import androidx.databinding.DataBindingUtil;
import com.google.android.material.bottomsheet.BottomSheetBehavior;
import com.google.android.material.bottomsheet.BottomSheetBehavior.BottomSheetCallback;
import com.spacex.ui.R;
import com.spacex.ui.about.AboutActivity;
import com.spacex.ui.databinding.RocketListBinding;
import com.spacex.ui.di.DaggerAppCompatActivityBase;
import com.spacex.ui.rockets.RocketsActivity;
import com.spacex.ui.welcome.WelcomeActivity;
import java.util.List;
import javax.inject.Inject;

public class RocketListActivity extends DaggerAppCompatActivityBase
    implements RocketListContract.View {

  @Inject RocketListContract.Presenter presenter;
  private RocketListAdapter adapter;
  private RocketListBinding viewBinding;
  private BottomSheetBehavior<ConstraintLayout> bottomSheetBehavior;

  public static void startActivity(@NonNull final Context context) {
    context.startActivity(new Intent(context, RocketListActivity.class));
  }

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    viewBinding = DataBindingUtil.setContentView(this, R.layout.rocket_list);
  }

  @Override
  public void setUpViews() {
    setUpToolbar();
    viewBinding.errorView.setOnRetryListener(() -> presenter.onRetry());
    viewBinding.recyclerView.setHasFixedSize(true);
    setUpSwipeRefresh();
    setUpBottomSheet();
  }

  private void setUpSwipeRefresh() {
    viewBinding.swipeRefreshLayout.setOnRefreshListener(() -> presenter.onRefresh());
    viewBinding.swipeRefreshLayout.setColorSchemeColors(
        ContextCompat.getColor(this, R.color.colorAccent));
  }

  private void setUpBottomSheet() {
    bottomSheetBehavior = BottomSheetBehavior.from(viewBinding.includeBottomSheet.bottomSheet);
    bottomSheetBehavior.setState(BottomSheetBehavior.STATE_HIDDEN);
    bottomSheetBehavior.setBottomSheetCallback(
        new BottomSheetCallback() {
          @Override
          public void onStateChanged(@NonNull final View view, final int state) {
            if (state == BottomSheetBehavior.STATE_HIDDEN) {
              presenter.onCloseFilterOptions();
            }
          }

          @Override
          public void onSlide(@NonNull final View view, final float v) {}
        });
    viewBinding.scrim.setOnClickListener(view -> presenter.onCloseFilterOptions());
    if (isAtLeastLollipop()) {
      viewBinding.scrim.setElevation(viewBinding.includeToolbar.toolbar.getElevation());
    }

    viewBinding.includeBottomSheet.filterAllRocketsRow.setOnClickListener(
        view -> presenter.onShowAllRocketsFilterOptionClick());

    viewBinding.includeBottomSheet.filterActiveRocketsRow.setOnClickListener(
        view -> presenter.onShowActiveRocketsFilterOptionClick());
  }

  private void setUpToolbar() {
    setSupportActionBar(viewBinding.includeToolbar.toolbar);
  }

  @Override
  public void showAsLoading() {
    viewBinding.errorView.setVisibility(View.GONE);
    viewBinding.emptyView.setVisibility(View.GONE);
    viewBinding.swipeRefreshLayout.setVisibility(View.GONE);
    viewBinding.includeProgressBar.progressBar.setVisibility(View.VISIBLE);
  }

  @Override
  public void showAsEmpty() {
    viewBinding.errorView.setVisibility(View.GONE);
    viewBinding.emptyView.setVisibility(View.VISIBLE);
    viewBinding.swipeRefreshLayout.setVisibility(View.GONE);
    viewBinding.includeProgressBar.progressBar.setVisibility(View.GONE);
  }

  @Override
  public void showAsErrorLoading() {
    viewBinding.emptyView.setVisibility(View.GONE);
    viewBinding.errorView.setVisibility(View.VISIBLE);
    viewBinding.swipeRefreshLayout.setVisibility(View.GONE);
    viewBinding.includeProgressBar.progressBar.setVisibility(View.GONE);
  }

  @Override
  public void showRockets(@NonNull final List<RocketUIM> rockets) {
    viewBinding.swipeRefreshLayout.setRefreshing(false);
    viewBinding.errorView.setVisibility(View.GONE);
    viewBinding.emptyView.setVisibility(View.GONE);
    viewBinding.swipeRefreshLayout.setVisibility(View.VISIBLE);
    viewBinding.includeProgressBar.progressBar.setVisibility(View.GONE);
    if (viewBinding.recyclerView.getAdapter() == null) {
      adapter = new RocketListAdapter(rockets, rocketId -> presenter.onRocketClicked(rocketId));
      viewBinding.recyclerView.setAdapter(adapter);
    } else {
      adapter.refreshData(rockets);
    }
  }

  @Override
  public void showAllRocketsFilterAsSelected() {
    viewBinding.includeBottomSheet.filterAllRocketsTick.setVisibility(View.VISIBLE);
    viewBinding.includeBottomSheet.filterActiveRocketsTick.setVisibility(View.GONE);
  }

  @Override
  public void showActiveRocketsFilterAsSelected() {
    viewBinding.includeBottomSheet.filterAllRocketsTick.setVisibility(View.GONE);
    viewBinding.includeBottomSheet.filterActiveRocketsTick.setVisibility(View.VISIBLE);
  }

  @Override
  public void navigateToRocketDetails(final int rocketId) {
    RocketsActivity.startActivity(this);
  }

  @Override
  public void navigateToWelcomeActivity() {
    WelcomeActivity.startActivity(this);
  }

  @Override
  public void navigateToAboutActivity() {
    AboutActivity.startActivity(this);
  }

  @Override
  public boolean onCreateOptionsMenu(final Menu menu) {
    getMenuInflater().inflate(R.menu.rocket_list, menu);
    return super.onCreateOptionsMenu(menu);
  }

  @Override
  public boolean onOptionsItemSelected(MenuItem item) {
    switch (item.getItemId()) {
      case R.id.filter_rockets:
        presenter.onFilterRocketsClick();
        return true;

      case R.id.show_welcome:
        presenter.onShowWelcomeClick();
        return true;

      case R.id.about:
        presenter.onAboutWelcomeClick();
        return true;
    }
    return super.onOptionsItemSelected(item);
  }

  @Override
  public void showFilterOptions() {
    showScrim();
    bottomSheetBehavior.setState(BottomSheetBehavior.STATE_EXPANDED);
  }

  @Override
  public void hideFilterOptions() {
    hideScrim();
    bottomSheetBehavior.setState(BottomSheetBehavior.STATE_HIDDEN);
  }

  private void showScrim() {
    viewBinding.scrim.setClickable(true);
    viewBinding
        .scrim
        .animate()
        .alpha(1f)
        .setDuration(this.getResources().getInteger(android.R.integer.config_shortAnimTime));
  }

  private void hideScrim() {
    viewBinding.scrim.setClickable(false);
    viewBinding
        .scrim
        .animate()
        .alpha(0f)
        .setDuration(this.getResources().getInteger(android.R.integer.config_shortAnimTime));
  }

  @Override
  public void onBackPressed() {
    presenter.onBackNavigationPressed();
  }

  @Override
  public void navigateBack() {
    super.onBackPressed();
  }
}

File: app/src/main/kotlin/com/spacex/ui/rocketList/RocketListActivity.kt
package com.spacex.ui.rocketList

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.core.content.ContextCompat
import androidx.databinding.DataBindingUtil
import com.google.android.material.bottomsheet.BottomSheetBehavior
import com.google.android.material.bottomsheet.BottomSheetBehavior.BottomSheetCallback
import com.spacex.ui.R
import com.spacex.ui.about.AboutActivity
import com.spacex.ui.common.isAtLeastLollipop
import com.spacex.ui.customViews.ErrorView
import com.spacex.ui.databinding.RocketListBinding
import com.spacex.ui.di.DaggerAppCompatActivityBase
import com.spacex.ui.rockets.RocketsActivity
import com.spacex.ui.thoughts.ThoughtsActivity
import com.spacex.ui.welcome.WelcomeActivity
import javax.inject.Inject

class RocketListActivity : DaggerAppCompatActivityBase(), RocketListContract.View {

    @Inject
    lateinit var presenter: RocketListContract.Presenter
    private var adapter: RocketListAdapter? = null
    private lateinit var viewBinding: RocketListBinding
    private lateinit var bottomSheetBehavior: BottomSheetBehavior<ConstraintLayout>

    companion object {

        fun startActivity(context: Context) {
            context.startActivity(Intent(context, RocketListActivity::class.java))
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        viewBinding = DataBindingUtil.setContentView(this, R.layout.rocket_list)
    }

    override fun setUpViews() {
        setUpToolbar()
        viewBinding.errorView.setOnRetryListener(object : ErrorView.OnRefreshListener {
            override fun onRefresh() {
                presenter.onRetry()
            }

        })
        viewBinding.recyclerView.setHasFixedSize(true)
        setUpSwipeRefresh()
        setUpBottomSheet()
    }

    private fun setUpToolbar() {
        setSupportActionBar(viewBinding.includeToolbar.toolbar)
    }

    private fun setUpSwipeRefresh() {
        viewBinding.swipeRefreshLayout.setOnRefreshListener { presenter.onRefresh() }
        viewBinding.swipeRefreshLayout.setColorSchemeColors(
                ContextCompat.getColor(this, R.color.colorAccent))
    }

    private fun setUpBottomSheet() {
        bottomSheetBehavior = BottomSheetBehavior.from(viewBinding.includeBottomSheet.bottomSheet)
        bottomSheetBehavior.state = BottomSheetBehavior.STATE_HIDDEN
        bottomSheetBehavior.setBottomSheetCallback(
                object : BottomSheetCallback() {
                    override fun onStateChanged(view: View, state: Int) {
                        if (state == BottomSheetBehavior.STATE_HIDDEN) {
                            presenter.onCloseFilterOptions()
                        }
                    }

                    override fun onSlide(view: View, v: Float) {}
                })
        viewBinding.scrim.setOnClickListener { presenter.onCloseFilterOptions() }
        if (isAtLeastLollipop()) {
            viewBinding.scrim.elevation = viewBinding.includeToolbar.toolbar.elevation
        }

        viewBinding.includeBottomSheet.filterAllRocketsRow.setOnClickListener { presenter.onShowAllRocketsFilterOptionClick() }
        viewBinding.includeBottomSheet.filterActiveRocketsRow.setOnClickListener { presenter.onShowActiveRocketsFilterOptionClick() }
    }

    override fun showAsLoading() {
        viewBinding.errorView.visibility = View.GONE
        viewBinding.emptyView.visibility = View.GONE
        viewBinding.swipeRefreshLayout.visibility = View.GONE
        viewBinding.includeProgressBar.progressBar.visibility = View.VISIBLE
    }

    override fun showAsEmpty() {
        viewBinding.errorView.visibility = View.GONE
        viewBinding.emptyView.visibility = View.VISIBLE
        viewBinding.swipeRefreshLayout.visibility = View.GONE
        viewBinding.includeProgressBar.progressBar.visibility = View.GONE
    }

    override fun showAsErrorLoading() {
        viewBinding.emptyView.visibility = View.GONE
        viewBinding.errorView.visibility = View.VISIBLE
        viewBinding.swipeRefreshLayout.visibility = View.GONE
        viewBinding.includeProgressBar.progressBar.visibility = View.GONE
    }

    override fun showRockets(rockets: List<RocketUIM>) {
        viewBinding.swipeRefreshLayout.isRefreshing = false
        viewBinding.errorView.visibility = View.GONE
        viewBinding.emptyView.visibility = View.GONE
        viewBinding.swipeRefreshLayout.visibility = View.VISIBLE
        viewBinding.includeProgressBar.progressBar.visibility = View.GONE
        if (viewBinding.recyclerView.adapter == null) {
            adapter = RocketListAdapter(rockets, object : RocketVH.OnRocketClickListener {
                override fun onClick(rocketId: String) {
                    presenter.onRocketClicked(rocketId)
                }

            })
            viewBinding.recyclerView.adapter = adapter
        } else {
            adapter?.refreshData(rockets)
        }
    }

    override fun showAllRocketsFilterAsSelected() {
        viewBinding.includeBottomSheet.filterAllRocketsTick.visibility = View.VISIBLE
        viewBinding.includeBottomSheet.filterActiveRocketsTick.visibility = View.GONE
    }

    override fun showActiveRocketsFilterAsSelected() {
        viewBinding.includeBottomSheet.filterAllRocketsTick.visibility = View.GONE
        viewBinding.includeBottomSheet.filterActiveRocketsTick.visibility = View.VISIBLE
    }

    override fun navigateToRocketDetails(rocketId: String, rocketIds: ArrayList<String>?) {
        RocketsActivity.startActivity(this, rocketIds, rocketId)
    }

    override fun navigateToWelcomeActivity() {
        WelcomeActivity.startActivity(this)
    }

    override fun navigateToThoughtsActivity() {
        ThoughtsActivity.startActivity(this)
    }

    override fun navigateToAboutActivity() {
        AboutActivity.startActivity(this)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.rocket_list, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.filter_rockets -> {
                presenter.onFilterRocketsClick()
                return true
            }

            R.id.show_welcome -> {
                presenter.onShowWelcomeClick()
                return true
            }

            R.id.thoughts -> {
                presenter.onThoughtsClick()
                return true
            }

            R.id.about -> {
                presenter.onAboutWelcomeClick()
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun showFilterOptions() {
        showScrim()
        bottomSheetBehavior.state = BottomSheetBehavior.STATE_EXPANDED
    }

    override fun hideFilterOptions() {
        hideScrim()
        bottomSheetBehavior.state = BottomSheetBehavior.STATE_HIDDEN
    }

    private fun showScrim() {
        viewBinding.scrim.isClickable = true
        viewBinding
                .scrim
                .animate()
                .alpha(1f).duration = this.resources.getInteger(android.R.integer.config_shortAnimTime).toLong()
    }

    private fun hideScrim() {
        viewBinding.scrim.isClickable = false
        viewBinding
                .scrim
                .animate()
                .alpha(0f).duration = this.resources.getInteger(android.R.integer.config_shortAnimTime).toLong()
    }

    override fun onBackPressed() {
        presenter.onBackNavigationPressed()
    }

    override fun navigateBack() {
        super.onBackPressed()
    }
}
